### ISSUE TITLE
VR: Calling subscribe two times without having handled the offer

### ISSUE THREAD

**Issue by JanFellner** [meetecho/janus-gateway#2955]
Looks like Janus is blocking that I can call subscribe a second time when i have not handled the offer properly.
I call subscribe get an ack but no response so my transaction times out.

I would expect to get a negative result or i would expect to get a new offer without having janus to wait for the initial offer<->answer.

The issue we have with it is that things *may* go wrong on the client side and in case i need to redo a subscription i currently see no other option beside tearing down the whole handle and reestablish everything.

(I notices this while implementing artificial error conditions in handshaking the subscription events with the client. E.g. what if the peerConnection on the client side is not accepting the offer for whatever reason or something times out and i simply need to recall subscribe?)

**Comment by lminiero** [comment]
Subscribe is an asynchronous transaction, and transactions don't timeout: at least, they definitely don't in Janus. The multistream VideoRoom _does_ allow you to do a new subscribe even when the previous offer/answer hasn't completed: in that case, it will "ignore" the request, meaning it will not send you a response, but will schedule a new offer for later on, as shown here:

https://github.com/meetecho/janus-gateway/blob/master/src/plugins/janus_videoroom.c#L9037-L9052

In case multiple subscribe requests arrive before an offer/answer has been finalized, the plugin may actually send you only a single new SDP offer instead of many, one that combines all the changes you made. This is expected and by design.

If by timeout you mean your own application layer is timing out because you don't get an actual event to the "subscribe" you sent, an easy fix may be sending a temporary event in the block I shared above: e.g., an "updating" event without an SDP offer, as that will only come later.

**Comment by JanFellner** [comment]
When communicating with janus i use a transaction based approach that waits on the websocket for the actual response on the request (based on the tranasaction id in the request.) So if i don´t get a reply within a certain amount of time i called it timed out.

The code you mention points out exactly what i am facing. I call subscribe but don´t get any reply on it as the previous request has not been handled on the client side.

The context with the sdp that comes later is unclear to me.

The event flow as i currently understand or use it is always like that:

1. Client -> Janus: Joining as subscriber (or subscribe if i am alread in the room), providing the mids i want to have
2. Client <- Janus: ACK
3. Client <- Janus: Structural information about the PeerConnection together with an SDP offer
4. Client -> Janus: start with the SDP answer
5. Client <- Janus: ACK
6. Client <- Janus: started

How to recover if between 3 and 4 something goes wrong?
I currently don´t see how to get back on track without dropping the hole plugin handle.

Also to mention from my side:
Our signalling server that sits next to janus already combines all requests that require an SDP renegotiation. So if janus also does that it hinders me here :). Any chance to add a subscribe property to no use that queuing inside janus? If i understand it correctly that is currently the logic that blocks the new sdp offer from janus as the old has not been handshaked yet...

**Comment by lminiero** [comment]
> Any chance to add a subscribe property to no use that queuing inside janus?

No. Queueing is needed to avoid glare. I'll just add the temporary event to make your client stack happy.

**Comment by JanFellner** [comment]
Lol, it´s not about the client stack :) I can live with any situation.

It´s actually how to extinguish the flame if something goes wrong. I can currently work with tearing down the plugin handle but i assume this is not the intended behavior for it.
How to enforce a new SDP if the last could not get handled (for whatever reason)

**Comment by lminiero** [comment]
Please let me now if this patch fixes it for you:

```
diff --git a/src/plugins/janus_videoroom.c b/src/plugins/janus_videoroom.c
index 23551f3b..56f9ba72 100644
--- a/src/plugins/janus_videoroom.c
+++ b/src/plugins/janus_videoroom.c
@@ -9039,6 +9039,13 @@ static void *janus_videoroom_handler(void *data) {
 					g_atomic_int_set(&subscriber->pending_offer, 1);
 					janus_mutex_unlock(&subscriber->streams_mutex);
 					JANUS_LOG(LOG_VERB, "Post-poning new offer, waiting for previous answer\n");
+					/* Send a temporary event */
+					event = json_object();
+					json_object_set_new(event, "videoroom", json_string("updating"));
+					json_object_set_new(event, "room", string_ids ?
+						json_string(subscriber->room_id_str) : json_integer(subscriber->room_id));
+					gateway->push_event(msg->handle, &janus_videoroom_plugin, msg->transaction, event, NULL);
+					json_decref(event);
 					/* Decrease the references we took before */
 					while(publishers) {
 						janus_videoroom_publisher *publisher = (janus_videoroom_publisher *)publishers->data;
@@ -9221,6 +9228,14 @@ static void *janus_videoroom_handler(void *data) {
 					g_atomic_int_set(&subscriber->pending_offer, 1);
 					janus_mutex_unlock(&subscriber->streams_mutex);
 					JANUS_LOG(LOG_VERB, "Post-poning new offer, waiting for previous answer\n");
+					/* Send a temporary event */
+					event = json_object();
+					json_object_set_new(event, "videoroom", json_string("updating"));
+					json_object_set_new(event, "room", string_ids ?
+						json_string(subscriber->room_id_str) : json_integer(subscriber->room_id));
+					gateway->push_event(msg->handle, &janus_videoroom_plugin, msg->transaction, event, NULL);
+					json_decref(event);
+					/* Decrease the references */
 					janus_refcount_decrease(&subscriber->ref);
 					janus_videoroom_message_free(msg);
 					continue;
```

**Comment by JanFellner** [comment]
That will just let the transaction time out earlier but not solve the actual problem (as is currently see it).

If something goes wrong (mobile connection, loose of data, loose of signalling, client is not accepting the sdp, whatever) how to enforce a new SDP from janus based on the requested streams/tracks?
How to recover if janus did provide an SDP but the answer on that is not received?

**Comment by lminiero** [comment]
> How to recover if janus did provide an SDP but the answer on that is not received?

The plugin will need an answer to finalize the offer/answer exchange, so without that there's no recovery possible. If it's the client rejecting the offer, then you can just send an artificial answer back yourself, e.g., with all m-lines set to inactive: depending on what the client didn't like, you can do something more drill down (e.g., only set the offending m-line to inactive).

I guess one possibility may be adding some sort of "reset" request to the plugin, where you tell act as if an answer was received even if it wasn't, but it would be a ugly hack with possibly several things that could go wrong. Besides, if the problem was something like losing connectivity, the PeerConnection is going to be invalid anyway, and so a new subscription should be restarted nevertheless.

**Comment by JanFellner** [comment]
Actually an "enforce" to the subscribe would be the flag i see there.

I would not want janus to cache anything here cause the external business logic already contains the queue to handle it accordingly.  (the client calls 1-n subscriptions and the requests are debounced, combined and then fired all together)

It´s okay for me that it caches the state for the offer but it should not necessarily have a queue to que up futher sdp exchanges in the future. (So as said in the beginning, the feature of the queing inside the videoroom plugin may not necessarily be something the client would want to use.)

Wouldn´t it be technically possible to overwrite the internal state in case another subscribe request comes in?  (I have no full picture of the state transitions there.)

**Comment by lminiero** [comment]
No, the queuing approach we use will not change. Besides, this is also tied to the offer/answer state in the Janus core, not only the plugin: if the plugin sends an offer, and you never send an answer back, it's first and foremost the Janus _core_ that will not like anything else than an answer. An answer to complete the cycle is needed there too (most importantly there).

**Comment by JanFellner** [comment]
I will put the amount of discarded plugin handles in a large waste bin and bring it to janus con this year ;]
And then we discuss about a cancel method *G*

An event that something has been queue when i try to do something (that is blocked by a pending answer) would help me now. 👍

Thanks for having time to answer...

**Comment by lminiero** [comment]
> An event that something has been queue when i try to do something (that is blocked by a pending answer) would help me now. +1

If you test the patch I posted a few messages above and let me know if it does the trick, I'll push it upstream :v:

**Comment by JanFellner** [comment]
Sadly i am currently no in the janus code side. I am building dockers and run them. Sololey working on the signalling server and frontend side :\. Maybe i find a spot next week.

**Comment by lminiero** [comment]
> is it guaranteed that it will receive an SDP offer later on with the previously missing information?

It should be, since if we mark it to be postponed, it means we're still handling a previous request, and are waiting for an answer. When we then do receive the answer, we check if there's a pending (previously postponed) subscription, and if so we generate and send a new offer. Notice that if more than one subscription was postponed while we were waiting for an answer, we'll only send a single new offer that reflects all the changes that were asked for in the meanwhile.

Now I don't know what's happening to you and if it's any different from what I described, what that's what the code is supposed to do. If you can summarize in a simple sequence diagram the exact steps that lead to a problem that never recovers (that isn't "client disconnects" as I believe in that case the subscription is invalid), then that would help traversing the code to see where the issue might be.

**Comment by lminiero** [comment]
About the "send new offer for postponed requests after we get the answer" part, this is the relevant code:

https://github.com/meetecho/janus-gateway/blob/master/src/plugins/janus_videoroom.c#L9972-L10010

**Comment by lminiero** [comment]
> Should the videoroom demo react in any way to this new `updating` event?

I don't think so. The videoroom demo doesn't track transactions for requests, it only reacts to events no matter what triggers them.

**Comment by flaviogrossi** [comment]
> > is it guaranteed that it will receive an SDP offer later on with the previously missing information?
> 
> It should be, since if we mark it to be postponed, it means we're still handling a previous request, and are waiting for an answer. 

> https://github.com/meetecho/janus-gateway/blob/master/src/plugins/janus_videoroom.c#L9972-L10010

I was simply surprised for not having an answer in that case to the subscribe request and i was wondering if it could happen for a subscriber to miss some publishers in this odd situation.
But looking at the code you posted, the client is guaranteed to have an `updated` event, since the `pending_offer` flag is set, so  there should be no problem.

Thanks for the clarification

**Comment by lminiero** [comment]
> Anyway, I'll push the patch to send the "updating" event for now, so that the client at least knows the message has been received.

Whoops, looks like it was pushed already, after I made a fix on something else yesterday :hand_over_mouth: 7d5a7515d741fca9bd2a0b2658c33d776b925241

**Comment by flaviogrossi** [comment]
> I guess you were not receiving the audio/video from the users marked in the second subscribe until much later, when something else triggered a new renegotiation?

Yes, that was exactly the scenario that i was worried about, and the reason why i asked for clarifications.
The end result was that everything was ok, but i don't know if it was because a later renegotiation (meaning it was just luck i was receiving the unlucky stream).

However for what i can see, i agree that from the code you sent it should not be a problem.

Isn't the offer you are looking for the `updated` event at `15:18:45.863`?
It is the websocket message exactly after the started response, and it contains an offer for `mid: "7"` and `feed_id: "71181825-ae69-46c4-aa1e-f42d9bdff714"`, which is coherent with the subscribe request with transaction `wKGPgMf8yLwO` which is not "answered" but only ack-ed

**Comment by lminiero** [comment]
Closing as not an issue.

**Comment by lminiero** [comment]
Thanks to both for the discussion and the feedback!
